### ISSUE TITLE
Tensorflow accesses .hdf5 weights file after it's been loaded

### ISSUE THREAD

**Issue by krafczyk** [tensorflow/tensorflow#54248]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux - Redhat 7.9
- TensorFlow installed from (source or binary): binary (conda)
- TensorFlow version (use command below): On all versions from 2.2 to 2.7 (2.7.0 for the latest)
- Python version: 3.8
- Bazel version (if compiling from source): 3.7.2
- GCC/Compiler version (if compiling from source): 10.2.0

**Describe the current behavior**

When loading a keras model from a temporary `.hdf5` weights file, the weights are read successfully read into the model, but Tensorflow outputs a DATA_LOSS warning after the `load_weights` method has finished executing indicating that Tensorflow is still trying to read the file.

**Describe the expected behavior**

I expect Tensorflow to not touch the weights file once the `load_weights` function as returned.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): no

**Standalone code to reproduce the issue**

```
import tensorflow as tf
import tensorflow_datasets as tfds
import tempfile
import os
import shutil

# Load MNIST Datasets
print("Load MNIST Dataset")
mnist_datasets, ds_info = tfds.load('mnist', as_supervised=True, with_info=True)
train_ds = mnist_datasets['train']
test_ds = mnist_datasets['test']

batch_size = 32

def normalize_img(image, label):
  """Normalizes images: `uint8` -> `float32`."""
  return tf.cast(image, tf.float32) / 255., label

try:
    autotune_opt = tf.data.AUTOTUNE
except:
    autotune_opt = tf.data.experimental.AUTOTUNE

train_ds = train_ds.map(
    normalize_img, num_parallel_calls=autotune_opt)
train_ds = train_ds.cache()
train_ds = train_ds.shuffle(ds_info.splits['train'].num_examples)
train_ds = train_ds.batch(batch_size)

test_ds = test_ds.map(
    normalize_img, num_parallel_calls=autotune_opt)
test_ds = test_ds.batch(batch_size)
test_ds = test_ds.cache()
test_ds = test_ds.prefetch(autotune_opt)

# Define function to create functional keras model
def create_keras_model(layer_dims):
    inp = tf.keras.layers.Input((28,28,1), name="Input_Name")
    last_layer = inp
    last_layer = tf.keras.layers.Reshape((28*28*1,), name="Input_Reshape")(last_layer)
    for i in range(len(layer_dims)):
        dim = layer_dims[i]
        last_layer = tf.keras.layers.Dense(dim, name=f"Layer_{i}")(last_layer)
    outp = tf.keras.layers.Dense(10, name="Output_Layer", activation='sigmoid')(last_layer)

    return tf.keras.Model(inputs=inp, outputs=outp)

# Create and train initial model:

print("Create and train initial keras model")
mdl_1 = create_keras_model([32,32,32])

mdl_1.compile(
    loss=tf.keras.losses.SparseCategoricalCrossentropy(),
    metrics=tf.keras.metrics.SparseCategoricalAccuracy())

mdl_1.fit(train_ds, epochs=1)

# Create function to test models
def eval_on_test(mdl):
    num_matches = 0
    total_examples = 0

    for X, Y in test_ds:
        # Eval model
        Y_eval = tf.argmax(mdl(X),axis=1).numpy()

        # Count
        total_examples += len(Y_eval)
        num_matches += (Y.numpy() == Y_eval).sum()

    return num_matches/total_examples

print("Initial model performance")
print(eval_on_test(mdl_1))

# Save weights to disk
mdl_weight_file = 'weights.hdf5'
mdl_1.save_weights(mdl_weight_file)

# Create new model and load from a temporary file
mdl_2 = create_keras_model([32,32,32])

with tempfile.NamedTemporaryFile('w+b') as temp_f:
    # Copy content to temporary file
    print("Copy weights to temp file")
    with open(mdl_weight_file, 'rb') as f:
        temp_f.write(f.read())

    # Flush to disk
    temp_f.flush()

    print("Load weights from temp file")
    # Load weights from temp file
    mdl_2.load_weights(temp_f.name)

print("temp file gone")

print("Loaded model performance")
print(eval_on_test(mdl_2))
```

**Other info / logs**

This the output I get by running this code locally:

```
Load MNIST Dataset
2022-02-02 15:45:23.299505: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2022-02-02 15:45:23.299702: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (iforge123): /proc/driver/nvidia/version does not exist
Create and train initial keras model
1875/1875 [==============================] - 6s 2ms/step - loss: 0.3753 - sparse_categorical_accuracy: 0.8929 
Initial model performance
0.9172
Copy weights to temp file
Load weights from temp file
2022-02-02 15:45:32.765376: W tensorflow/core/util/tensor_slice_reader.cc:96] Could not open /tmp/tmpfpsuz_25: DATA_LOSS: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?
temp file gone
Loaded model performance
0.9172
```

I tried running this code on colab, however I can't see any of the tensorflow logging lines I see on my local machine.

**Comment by krafczyk** [comment]
This might be expected behavior, but then I'd like a way to determine that tensorflow is finished reading the file before releasing the temporary file.

**Comment by tilakrayal** [comment]
@krafczyk ,
I was able to execute the mentioned code with out any Data loss warnings/issues.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/4f18c11a8f4c2d065dbb0bc27dce8e10/untitled211.ipynb).

**Comment by krafczyk** [comment]
@tilakrayal I worked around the colab logging issue by first writing the script to a file, and then using `!` magic commands to execute the script. If you know a proper way to show logging on colab please let me know.

Here is my [gist](https://colab.research.google.com/drive/1JE6NuBB2oJ_0MMiEcomP4kfXJOidp9Pd?usp=sharing)

**Comment by tilakrayal** [comment]
@chunduriv ,
I was able to reproduce the issue in tf v2.5, v2.7 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/56e11b6f728e3449c61a90a9c1a6456a/untitled214.ipynb).

**Comment by chunduriv** [comment]
@krafczyk,

Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999).Thanks!

**Comment by krafczyk** [comment]
@chunduriv Sure. Since the error message is originating from the tensorflow C library, i figured this would be a better place to start.

**Comment by krafczyk** [comment]
I've opened the issue at keras-team/keras here: keras-team/keras#16048

**Comment by chunduriv** [comment]
@krafczyk, Can you please close this issue, since it is tracked [there](https://github.com/keras-team/keras/issues/16048). Thanks!
